File: matlab/bin_spike_4_PEV_cur_and_othr_tg_1stTrlExcd_Pfct.m
function [binspikes curr_targ_idx other_targ_idx] = bin_spike_4_PEV_cur_and_othr_tg_1stTrlExcd_Pfct(...   %location last touched 
    t_spike,... %spike time stamps
    t_trig,...  %event time stamps
    TrialID,... %behavioral event codes (eg correct/incorrect)
    width,...   % bin width
    slide,...
    Fx_ana_win,...
    Go_ana_win,...
    Fb_ana_win,...
    numBx,...  % number of choices in the task
    Cycle,...   % cycle to analyze
    targnum,... % number of targets in the task
    touchOrder)



binspikes = [];
curr_targ_idx = [];
other_targ_idx = [];

w = 30; % Gaussian SD not used
Gauss_width = max([11 6*w+1]); % 
kernel      = normpdf(-floor(Gauss_width/2):floor(Gauss_width/2),0,w);

COI = Cycle;



%%-------------------------------------------------------------------------
% Extract trial event data with Reveal event from raw data (trials appropriate for analyses)
RevealTime = t_trig(2:end,5);
for iii = 1:length(RevealTime)
    if ~isempty(RevealTime{iii}) && length(RevealTime{iii}) == 1
    ActiveTrial(iii) = ~isnan(RevealTime{iii});
    else
        ActiveTrial(iii) = 0;
    end
end
% ?s‚²‚Æ‚ÌTaskí—Ş
cellTask = TrialID(2:end,1);
trialcounts = length(cellTask);
cellTargnum = TrialID(2:end,3);
cellChoicenum = TrialID(2:end,4);
cellCyclenum = TrialID(2:end,5);
cellTouchID = TrialID(2:end,8);

celltcode = TrialID(2:end,6);
for iii = 1:length(celltcode)
    if isempty(celltcode{iii})
        ActiveTrial(iii) = 0;
        celltcode{iii} = NaN;
    end
end


TouchCountInCycle = [];
if cellTouchID{1} == 1
    cellCorCount(1) = 1;
    cellCorrectCount = 1;
    TouchCountInCycle(1) = 1;
else
    cellCorCount(1) = NaN;
    cellCorrectCount = 0;
    TouchCountInCycle(1) = 1;
end

lastCycle = 1;

for i = 2:trialcounts
    comp = 0; % comparison ?
    if isnan(cellCyclenum{i-1})
        comp = lastCycle; % = 1
    else
        comp = cellCyclenum{i-1};
        if ~isnan(cellCyclenum{i})
            lastCycle = cellCyclenum{i};
        end
    end
    if cellCyclenum{i} == comp | isnan(cellCyclenum{i})% cycle continued
        if ~isnan(celltcode{i})
        TouchCountInCycle(i) = TouchCountInCycle(i-1) + 1;
        else     
        TouchCountInCycle(i) = TouchCountInCycle(i-1); 
        end
         
        if cellTouchID{i} == 1
            cellCorrectCount = cellCorrectCount + 1; % 1st or 2nd correct touch ?
            cellCorCount(i) = cellCorrectCount;
        else
            cellCorCount(i) = cellCorrectCount; %NaN;
        end
    else % new cycle
        if ~isnan(celltcode{i})
            TouchCountInCycle(i) = 1;
        else
            TouchCountInCycle(i) = 0;
        end
       
        cellCorrectCount = 0;
        if cellTouchID{i} == 1
            cellCorrectCount = cellCorrectCount + 1;
            cellCorCount(i) = cellCorrectCount;
        else
            cellCorCount(i) = cellCorrectCount; %NaN
        end
    end
end





celllevlatency = TrialID(2:end,10);
for iii = 1:length(celllevlatency)
    if isempty(celllevlatency{iii})
        ActiveTrial(iii) = 0;
        celllevlatency{iii} = NaN;
    end
end

celltouchlatency = TrialID(2:end,11);
for iii = 1:length(celltouchlatency)
    if isempty(celltouchlatency{iii})
        ActiveTrial(iii) = 0;
        celltouchlatency{iii} = NaN;
    end
end

cellTargLoc = TrialID(2:end,2);

cellOtherTarg = {};

nTrl2OtherTargTch = [];

for i_trl = 1:length(cellTargLoc) % for all trials
    if ~isempty(cellTouchID{i_trl}) && cellTouchID{i_trl} == 1 &&  cellTargnum{i_trl} == 2 % correct touch in the 2-target condition
        currentTargs = [];
        currentTargs = cellTargLoc{i_trl}; % indices of current 2 targets
       
        cellOtherTarg{i_trl} = currentTargs(find(currentTargs ~= celltcode{i_trl}));
        
        % find if this is a perfect cycle
        ntrl_to_check = [];
%         nTrl2OtherTargTch = 99;
        if cellCorCount(i_trl) == 1 % first touch
            ntrl_to_check = length(celltcode(i_trl:end)) -1;
            for i_check = 1:ntrl_to_check
                if ( celltcode{i_trl + i_check} == cellOtherTarg{i_trl})
                    nTrl2OtherTargTch(i_trl) = i_check;
                    break;
                end
            end
        elseif cellCorCount(i_trl) == 2
            ntrl_to_check = length(celltcode(1:i_trl)) -1;
            for i_check = 1:ntrl_to_check
                if celltcode{i_trl - i_check} == cellOtherTarg{i_trl}              
                    nTrl2OtherTargTch(i_trl) = i_check;
                    break;
                end
            end
            
        end
        
    else
        cellOtherTarg{i_trl} = NaN;  % all other touches (ignored in the 'other target' analysis)
    end
       
end

%% ------------------------------------------------------------------------
trialnum = size(TrialID, 1);
trialnum = trialnum - 1;
    
sps.Fx_bin = [];
sps.Go_bin = [];
sps.Fb_bin = [];
sps.All_bin = [];
%% ------------------------------------------------------------------------

trial_count = 0;
excld = 99;

                
for i = 2:trialnum
    if   ActiveTrial(i)  == 1 && ...                 % Trial n had  Reveal event
         ((cellCyclenum{i} == 1 && TouchCountInCycle(i) > 1) || (cellCyclenum{i} > 1)) && ...
         ~isempty(find(cellCyclenum{i} == COI)) && ...  % Trial n belongs to cycle of interest
         cellCorCount(i) == touchOrder && ...     % ordinal number is this correct touch in a given cycle?
         cellTouchID{i}  == 1 && ...              % Is this touch target-touch (1) or non-target touch (2)?
         cellTask{i}     == 1 && ....             % task (1->spatial; 2->object)
         cellChoicenum{i} == numBx && ...         % Trial n has numBx number of choice (ie 5 or 6 choice)
         cellTargnum{i}  == targnum && ...        % target condition (1 target / 2 target)
         celltcode{i} ~= excld && ...
         ~isnan(t_trig{i+1, 6}) && ...
         t_trig{i+1, 1} +  Fx_ana_win(1) > 0&& ...
         length(nTrl2OtherTargTch) >= i && ...
         (cellCyclenum{i} >= 2 && nTrl2OtherTargTch(i) == 1 && TouchCountInCycle(i) == 1)
  
        trial_count = trial_count + 1;
        

       %% Extracting fixation period activity in 'trial n'
        % set time of data (t_data0 -> t_data1)
        t_range = Fx_ana_win(2)-Fx_ana_win(1)+1;  
        bn = (t_range-width)/slide+1;     % number of sliding bins
        
        t_trig_data = t_trig{i+1,1}; % trigger is 1 (fixation onset)
        t_data0 = t_trig_data + Fx_ana_win(1);
        t_data1 = t_trig_data + Fx_ana_win(2);
        u_data = zeros(1,t_data1-t_data0+1);
        u_data =  t_spike{i}(t_data0:t_data1);
        % Gaussian Kernel Convolution
        % dummy       = conv(u_data,kernel); % pre allocation for speed-up
        % u_data      = dummy(floor(Gauss_width/2)+1:end-floor(Gauss_width/2)); % mode of Gaussian centered on spike -> noncausal      
        for j=1:bn
            bs = slide*(j-1)+1;
            be = bs+width-1;            
            u_bin = zeros(1,width);
            u_bin = u_data(1,bs:be);
            sps.Fx_bin(trial_count,j) = length(find(u_bin==1)) / (width/1000);  
        end
        
       %% Extracting Go period activity in 'trial n'
        % set time of data (t_data0 -> t_data1)
        t_range = Go_ana_win(2)-Go_ana_win(1)+1;  
        bn = (t_range-width)/slide+1;     % number of sliding bins      
        
        t_trig_data = t_trig{i+1,3}; % trigger is 3 (lever-release onset)       
        t_data0 = t_trig_data + Go_ana_win(1);
        t_data1 = t_trig_data + Go_ana_win(2);
        u_data = zeros(1,t_data1-t_data0+1);
        u_data =  t_spike{i}(t_data0:t_data1);
        % Gaussian Kernel Convolution
        % dummy       = conv(u_data,kernel); % pre allocation for speed-up
        % u_data      = dummy(floor(Gauss_width/2)+1:end-floor(Gauss_width/2)); % mode of Gaussian centered on spike -> noncausal      
        for j=1:bn
            bs = slide*(j-1)+1;
            be = bs+width-1;            
            u_bin = zeros(1,width);
            u_bin = u_data(1,bs:be);
            sps.Go_bin (trial_count,j) = length(find(u_bin==1)) / (width/1000);  %sum(u_bin); % length(find(u_bin==1));
        end
       
       %% Extracting Fb period activity in 'trial n'
        % set time of data (t_data0 -> t_data1)
        t_range = Fb_ana_win(2)-Fb_ana_win(1)+1;  
        bn = (t_range-width)/slide+1;     % number of sliding bins      
        
        t_trig_data = t_trig{i+1,5}; % trigger is 5 (feedback onset)  
        t_data0 = t_trig_data + Fb_ana_win(1);
        t_data1 = t_trig_data + Fb_ana_win(2);
        u_data = zeros(1,t_data1-t_data0+1);
        u_data =  t_spike{i}(t_data0:t_data1);
        % Gaussian Kernel Convolution
        % dummy       = conv(u_data,kernel); % pre allocation for speed-up
        % u_data      = dummy(floor(Gauss_width/2)+1:end-floor(Gauss_width/2)); % mode of Gaussian centered on spike -> noncausal      
        for j=1:bn
            bs = slide*(j-1)+1;
            be = bs+width-1;            
            u_bin = zeros(1,width);
            u_bin = u_data(1,bs:be);
            sps.Fb_bin(trial_count,j) = length(find(u_bin==1)) / (width/1000);  %sum(u_bin); % length(find(u_bin==1));
        end
        
       
        
      
        % identifing location of target (target idx)
        curr_targ_idx(trial_count) = celltcode{i};
        other_targ_idx(trial_count) = cellOtherTarg{i};
        
    else
        trial_count = trial_count;
    end
end


%% ------------------------------------------------------------------------
% Creating unified spike data
sps.All_bin = horzcat(sps.Fx_bin, sps.Go_bin, sps.Fb_bin);
if trial_count > 0
    binspikes = sps.All_bin;
else
    binspikes = [];
end

File: python/bin_spike_4_PEV_cur_and_othr_tg_1stTrlExcd_Pfct.py
import numpy as np
from scipy.stats import norm

def bin_spike_4_PEV_cur_and_othr_tg_1stTrlExcd_Pfct(t_spike, t_trig, TrialID, width, slide, Fx_ana_win, Go_ana_win, Fb_ana_win, numBx, Cycle, targnum, touchOrder):
    binspikes = []
    curr_targ_idx = []
    other_targ_idx = []

    w = 30  # Gaussian SD not used
    Gauss_width = max(11, 6 * w + 1)  # 
    kernel = norm.pdf(np.arange(-np.floor(Gauss_width / 2), np.floor(Gauss_width / 2) + 1), 0, w)

    COI = Cycle

    # Extract trial event data with Reveal event from raw data (trials appropriate for analyses)
    RevealTime = t_trig[1:, 5]
    ActiveTrial = np.array([0 if len(rt) == 0 or len(rt) != 1 else not np.isnan(rt) for rt in RevealTime])

    # Task specifics
    cellTask = TrialID[1:, 0]
    trialcounts = len(cellTask)
    cellTargnum = TrialID[1:, 2]
    cellChoicenum = TrialID[1:, 3]
    cellCyclenum = TrialID[1:, 4]
    cellTouchID = TrialID[1:, 7]
    celltcode = TrialID[1:, 5]

    for iii in range(len(celltcode)):
        if not celltcode[iii]:
            ActiveTrial[iii] = 0
            celltcode[iii] = np.nan

    TouchCountInCycle = []
    cellCorCount = np.full(trialcounts, np.nan)
    cellCorrectCount = 0
    TouchCountInCycle.append(1 if cellTouchID[0] == 1 else 0)

    lastCycle = 1
    for i in range(1, trialcounts):
        comp = lastCycle if np.isnan(cellCyclenum[i-1]) else cellCyclenum[i-1]
        if not np.isnan(cellCyclenum[i]):
            lastCycle = cellCyclenum[i]
        if cellCyclenum[i] == comp or np.isnan(cellCyclenum[i]):  # cycle continued
            TouchCountInCycle.append(TouchCountInCycle[i-1] + 1 if cellTouchID[i] else TouchCountInCycle[i-1])
            if cellTouchID[i] == 1:
                cellCorrectCount += 1
            cellCorCount[i] = cellCorrectCount
        else:  # new cycle
            TouchCountInCycle.append(1 if not np.isnan(celltcode[i]) else 0)
            cellCorrectCount = 0
            if cellTouchID[i] == 1:
                cellCorrectCount += 1
            cellCorCount[i] = cellCorrectCount

    celllevlatency = TrialID[1:, 9]
    celltouchlatency = TrialID[1:, 10]
    cellTargLoc = TrialID[1:, 1]

    cellOtherTarg = []
    nTrl2OtherTargTch = np.full(trialcounts, np.nan)

    for i_trl in range(len(cellTargLoc)):
        if cellTouchID[i_trl] == 1 and cellTargnum[i_trl] == 2:  # correct touch in 2-target condition
            currentTargs = np.array(cellTargLoc[i_trl])
            cellOtherTarg.append(currentTargs[currentTargs != celltcode[i_trl]])

            ntrl_to_check = []
            if cellCorCount[i_trl] == 1:  # first touch
                ntrl_to_check = len(celltcode[i_trl:]) - 1
                for i_check in range(ntrl_to_check):
                    if celltcode[i_trl + i_check] == cellOtherTarg[i_trl]:
                        nTrl2OtherTargTch[i_trl] = i_check
                        break
            elif cellCorCount[i_trl] == 2:
                ntrl_to_check = len(celltcode[:i_trl]) - 1
                for i_check in range(ntrl_to_check):
                    if celltcode[i_trl - i_check] == cellOtherTarg[i_trl]:
                        nTrl2OtherTargTch[i_trl] = i_check
                        break
        else:
            cellOtherTarg.append(np.nan)

    # Define sps for storing spike data
    sps = {'Fx_bin': [], 'Go_bin': [], 'Fb_bin': [], 'All_bin': []}

    trial_count = 0
    excld = 99

    trialnum = trialcounts

    for i in range(1, trialnum):
        if ActiveTrial[i] == 1 and \
            ((cellCyclenum[i] == 1 and TouchCountInCycle[i] > 1) or (cellCyclenum[i] > 1)) and \
            not np.isnan(np.where(cellCyclenum[i] == COI)) and \
            cellCorCount[i] == touchOrder and \
            cellTouchID[i] == 1 and \
            cellTask[i] == 1 and \
            cellChoicenum[i] == numBx and \
            cellTargnum[i] == targnum and \
            celltcode[i] != excld and \
            not np.isnan(t_trig[i+1, 5]) and \
            t_trig[i+1, 0] + Fx_ana_win[0] > 0 and \
            len(nTrl2OtherTargTch) >= i and \
            (cellCyclenum[i] >= 2 and nTrl2OtherTargTch[i] == 1 and TouchCountInCycle[i] == 1):

            trial_count += 1

            # Extracting fixation period activity in 'trial n'
            t_range = Fx_ana_win[1] - Fx_ana_win[0] + 1
            bn = int((t_range - width) / slide + 1)

            t_trig_data = t_trig[i + 1, 0]
            t_data0 = t_trig_data + Fx_ana_win[0]
            t_data1 = t_trig_data + Fx_ana_win[1]
            u_data = np.zeros(t_data1 - t_data0 + 1)
            u_data = t_spike[i][t_data0:t_data1]

            for j in range(bn):
                bs = slide * (j - 1) + 1
                be = bs + width - 1
                u_bin = np.zeros(width)
                u_bin = u_data[bs:be]
                sps['Fx_bin'].append(np.sum(u_bin == 1) / (width / 1000))

            # Extracting Go period activity in 'trial n'
            t_range = Go_ana_win[1] - Go_ana_win[0] + 1
            bn = int((t_range - width) / slide + 1)

            t_trig_data = t_trig[i + 1, 2]
            t_data0 = t_trig_data + Go_ana_win[0]
            t_data1 = t_trig_data + Go_ana_win[1]
            u_data = np.zeros(t_data1 - t_data0 + 1)
            u_data = t_spike[i][t_data0:t_data1]

            for j in range(bn):
                bs = slide * (j - 1) + 1
                be = bs + width - 1
                u_bin = np.zeros(width)
                u_bin = u_data[bs:be]
                sps['Go_bin'].append(np.sum(u_bin == 1) / (width / 1000))

            # Extracting Fb period activity in 'trial n'
            t_range = Fb_ana_win[1] - Fb_ana_win[0] + 1
            bn = int((t_range - width) / slide + 1)

            t_trig_data = t_trig[i + 1, 4]
            t_data0 = t_trig_data + Fb_ana_win[0]
            t_data1 = t_trig_data + Fb_ana_win[1]
            u_data = np.zeros(t_data1 - t_data0 + 1)
            u_data = t_spike[i][t_data0:t_data1]

            for j in range(bn):
                bs = slide * (j - 1) + 1
                be = bs + width - 1
                u_bin = np.zeros(width)
                u_bin = u_data[bs:be]
                sps['Fb_bin'].append(np.sum(u_bin == 1) / (width / 1000))

            # Identifying location of target (target idx)
            curr_targ_idx.append(celltcode[i])
            other_targ_idx.append(cellOtherTarg[i])

    # Creating unified spike data
    sps['All_bin'] = np.hstack([sps['Fx_bin'], sps['Go_bin'], sps['Fb_bin']])

    return binspikes, curr_targ_idx, other_targ_idx